DM-28035: Backfill tool for DBB endpoint manager (Stage 1)

mxk62 opened the pull request:
Added a backfill tool allowing one to populate endpoint manager's table with entries related to files which were already in the storage area before its initial deployment.

Note that this tool does *not* check if a file is ingested to the Butler dataset repository managed by the endpoint manager nor it tries to ingest the file if it isn't. It merely makes the endpoint manager "aware of" these files.

mxk62 committed:
Move get_checksum() to utils.py module

mxk62 committed:
Move out file discovery methods to separate module

Methods responsible for the file discovery resided in Finder's module as
it was the only manager's component using them.  However, the backfill
tool will most likely be using one of them as well so I placed them in a
separate module.

mxk62 committed:
Add a backfill command

DBB endpoint buffer manager didn't have the capability to backfill
database tables it uses with entries regarding files which already
existed in the storage area before it was deployed.  To address this
limitation I added a new command, backfill, which allow an operator to
populate the file and event tables with entries related to these files.

At the moment, the command does *not* verify if the files are really
ingested to a Butler dataset repository.  It merely makes the manager
"aware of" of these files by making an entries in file and event tables.

mxk62 committed:
Fix some docstrings

mxk62 committed:
Define __all__'s as lists consistently

The lists of public objects of the modules, __all__, was a list in some
modules or a tuple in others.  Following official Python documentation, I
redefined each existing __all__ as a list.

mxk62 committed:
Make a minor refactoring of ingest command

mxk62 committed:
Describe backfill tool in documentation

Add the description of the backfill tool usage in the documentation and
included and example configuration file for it.

mxk62 committed:
Fix configuration example in the documentation

The indentation of one section in an example configuration file
described in the documentation was not correct.  Fixed it.

mxk62 committed:
Fix example configuration for Finder

The example configuration file for Finder had an error.  Fixed it.

mxk62 committed:
Get rid of id query

The backfill tool was making an explicit database query to find out the
id of the file record it added in the earlier step.  I replace it
SQLAlechmy's session.flush() which pushes session changes to the
database's transaction buffer and allow to get id of the added file
record without making a separate query.

mxk62 committed:
Issue a short summary after backfilling finishes

Modify the backfill tool to make it issue a short summary of the run.

mxk62 committed:
Add a validation schema for backfill tool

mxk62 committed:
Fix errors in validation schemas and sample files

After adding a validation schema for the backfill tool I noticed a bunch
of small errors in the schemas for the Finder and the Ingester as well
as in the example configuraion files.  Fixed all of them to make sure
that the example files can be properly validated.

mxk62 committed:
Add unit tests for validation

Added unit tests which verifies that sample configuration files validate
against the schemas.

MichelleGower commented on doc/lsst.dbb.buffmngrs.endpoint/get-started.rst:
```diff
@@ -269,13 +269,80 @@ to make another ingest attempt for selected files.
+Backfill database records
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Most likely some files were already present in the storage area and ingested to
```
Most likely -> In the case where some files ...

MichelleGower commented on doc/lsst.dbb.buffmngrs.endpoint/get-started.rst:
```diff
@@ -269,13 +269,80 @@ to make another ingest attempt for selected files.
+
+.. warning::
+
+   At the moment, the backfill tool does not verify if any of the files found
```
Remove "At the moment"

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py:
```diff
@@ -0,0 +1,190 @@
+
+        self.session = config["session"]
+
+        required = {"event", "file"}
```
Move shared code to function or at least comment.  It took me a minute to figure out why this second block of code existed because it looks so much similar to the previous block.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py:
```diff
@@ -0,0 +1,139 @@
+def backfill():
+    """Populate the database with entries for historical files.
+    """
+    pass
```
I don't know enough about click groups.  Just verifying that this method is necessary since it only contains pass or is the pass line needed?

mxk62 replied:
As far as my understanding of Click goes, yes, it is necessary to nest commands.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py:
```diff
@@ -0,0 +1,190 @@
+        acc = dict(failure=0, success=0)
+        sources = [glob(os.path.join(self.storage, s)) for s in self.sources]
+        for source in chain(*sources):
+            logger.debug(f"{source}: setting as file source")
```
Use lazy % formatting in logging functions (https://developer.lsst.io/stack/logging.html#basic-usage-in-python but it seems to be LSST logging specific.)  If want to make a separate ticket to make this change across entire code, feel free to postpone.

mxk62 replied:
None of the manager's commands use lazy formatting at the moment. I'd like to keep things consistent so I'll make a separate ticket for that work.

As a side note, all f-strings used by log messages include basic Python data structures only so the overhead resulting from eager formatting should be negligible.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py:
```diff
@@ -0,0 +1,190 @@
+                dirname, filename = os.path.split(relpath)
+                logger.debug(f"{relpath}: starting processing")
+
+                logger.debug(f"checking if already tracked")
```
f-string without any variables (i.e., remove the _f_)  Multiple lines are this way.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/search.py:
```diff
@@ -0,0 +1,209 @@
+
+
+@register
+def scan(directory, blacklist=None, **kwargs):
```
kwargs not used (or if really want to keep, missing in docstring)

mxk62 replied:
Yes, I want to keep it to ensure that search methods can be used interchangeably. I will document it in the docstring

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py:
```diff
@@ -0,0 +1,139 @@
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+"""Command line interface for the Backfill component.
+"""
+import click
```
click is a third-party import and should go later than standard imports (e.g. imports of importlib, inspect, and logging).

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py:
```diff
@@ -0,0 +1,139 @@
+    logger.info("setting up database connection...")
+    config = configuration["database"]
+
+    module = importlib.import_module("sqlalchemy.pool")
```
Should this be using LSST's doImport?

mxk62 replied:
Nothing in the _framework_ depends on LSST Stack software. Only GenX Butler ingest plugins use LSST code. I'd prefer to keep it this way in case plugins will end up in a separate package.

MichelleGower commented on doc/lsst.dbb.buffmngrs.endpoint/get-started.rst:
```diff
@@ -269,13 +269,80 @@ to make another ingest attempt for selected files.
+a Butler dataset repository before the deployment of the DBB endpoint buffer
+manager.  
+
+Use ``backfill`` tool to populate the endpoint manager's database tables with
```
tool -> command or subcommand as went with subcommands instead of separate script?

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py:
```diff
@@ -0,0 +1,190 @@
+logger = logging.getLogger(__name__)
+
+
+class Backfill(object):
```
object isn't needed.  In Python 3 inheritance from object is implicit.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py:
```diff
@@ -0,0 +1,139 @@
+              help="Validate configuration before starting the service.")
+@click.argument("filename", type=click.Path(exists=True))
+def start(filename, dump, validate):
+    """Starts a finder using a configuration from FILENAME.
```
"Starts a **finder**"?  Is this a copy/paste mistake or have I misunderstood something?

mxk62 replied:
Yes, it's a copy/paste mistake. Will fix.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py:
```diff
@@ -0,0 +1,139 @@
+def start(filename, dump, validate):
+    """Starts a finder using a configuration from FILENAME.
+    """
+    with open(filename) as f:
```
A bunch of the start function is the same across all three */commands.py files.  The shared code should be moved to function(s).  There are blocks of code that could be smaller functions (e.g., setting up the sqlalchemy pool) or might be able keep as reusable larger function.  Again, ok if want to postpone to a new ticket.

mxk62 replied:
I'll factor our repeated code to functions.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/validation.py:
```diff
@@ -286,3 +289,93 @@
     - ingester
 """
+
+backfill = """
```
Constants should be UPPER_CASE?

MichelleGower commented on doc/lsst.dbb.buffmngrs.endpoint/get-started.rst:
```diff
@@ -269,13 +269,80 @@ to make another ingest attempt for selected files.
+manager.  
+
+Use ``backfill`` tool to populate the endpoint manager's database tables with
+entires regarding these files.
```
typo: entires -> entries?

MichelleGower commented on doc/lsst.dbb.buffmngrs.endpoint/get-started.rst:
```diff
@@ -269,13 +269,80 @@ to make another ingest attempt for selected files.
+
+      You can specify multiple sources.  The source entry can be either a file
+      or a directory.  You can use unix style pathname pattern expansion
+      when specyfing source's name (tilde expansion is not supported though).
```
typo: specyfing -> specifying

MichelleGower commented on doc/lsst.dbb.buffmngrs.endpoint/get-started.rst:
```diff
@@ -269,13 +269,80 @@ to make another ingest attempt for selected files.
+   .. note::
+
+      To prevent Python YAML parser from accidental conversion of strings like
+      ``2021-01-01`` into a ``datatime`` objects, enclose them in quotes.
```
typo: datatime -> datetime

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py:
```diff
@@ -0,0 +1,190 @@
+        total = sum(x for x in acc.values())
+        if total == 0:
+            logger.warning("no backfill attempts were made")
+        if acc["failure"] != 0:
```
Not sure why this isn't an elif (i.e., if no backfill **attempts** were made can one have failures or successes?).

mxk62 replied:
It was rectified while addressing the comment below, the one about making failures more granular.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py:
```diff
@@ -0,0 +1,190 @@
+    def run(self):
+        """Start the backfill process.
+        """
+        acc = dict(failure=0, success=0)
```
Seems to be keeping track of counts as opposed to the more generic accumulating values.  Code would be easier to read if variable name was some form of the word counts.

MichelleGower replied:
Also after reading farther down the code, failure covers many different failure mechanisms including already tracked.  It would be helpful if these counts were separated into more categories, especially those that are unusual internal issues (sqlalchemy outage, file disappeared in the middle of processing, etc) vs already in DB (which should just be an informational count so one can rerun backfill on the same directories without it reporting already tracked as failures).

mxk62 replied:
You're totally right. I will:
1. change "acc" to "counts",
2. introduce new counter "tracked",
3. decrease the severity of the log message about already track file from WARNING to INFO.

MichelleGower commented on doc/lsst.dbb.buffmngrs.endpoint/get-started.rst:
```diff
@@ -269,13 +269,80 @@ to make another ingest attempt for selected files.
+      You can specify multiple sources.  The source entry can be either a file
+      or a directory.  You can use unix style pathname pattern expansion
+      when specyfing source's name (tilde expansion is not supported though).
+      All sources are assumed to be relative with regard to the storage area.
```
If abs paths that include the storage path are given does it work?  If it breaks, change the wording to "must be" and consider checking not abs paths.  If it works and are ok with someone using it, also change the wording.  Code definitely should not support absolute paths that do not include the storage path.

mxk62 replied:
I will change the wording in the documentation and add a guard against absolute paths.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py:
```diff
@@ -0,0 +1,190 @@
+                acc["failure"] += 1
+                continue
+
+            paths = scan(source, **self.search_opts) if os.path.isdir(source) \
```
This section of code is really hard to follow.   Please add more comments.  Also feels like there is a bunch of path manipulations that could be cleaned up (e.g., adding, removing, adding, removing).  Seems like there should be some way of removing the self.storage from the source and then only add the additional bits to the path in the loop.

mxk62 replied:
I will add comments which hopefully make the code easier to follow.

However, I don't really see how path manipulations may be simplified if we would like to keep current flexibility in specification of the file sources.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py:
```diff
@@ -0,0 +1,190 @@
+        if acc["failure"] != 0:
+            logger.warning("some backfill attempts failed")
+        else:
+            logger.info("all found files were backfilled successfully")
```
Since the code has the count of successes why not print it?

MichelleGower replied:
Also referring back to the earlier comment about "already tracked" shouldn't be an error (i.e., I should be able to run and rerun backfill on the same directory and it should only be errors if it can't backfill files that need to be backfilled), a count of the "already tracked" could be printed here as info (or warning).

mxk62 replied:
> Since the code has the count of successes why not print it?

No particular reason, I just assumed no one would care how many files were backfilled successfully _if_ there were no failures (of any kind).

However, in the context your earlier suggestion about distinguishing between different types of failed backfills (DB errors vs. already tracked), now it will make sense to presents these statistics. I will make the necessary changes.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py:
```diff
@@ -0,0 +1,139 @@
+
+    logger.info("starting Backfill...")
+    component = Backfill(backfill_config)
+    component.run()
```
Not exactly sure what line to add this comment with, but if there are backfill failures/errors (not counting "already tracked" as errors), shouldn't endmgr exit with a non-zero exit code?

mxk62 replied:
I'll make the code to raise a RuntimeError exception if any errors were encountered during the backfilling.

MichelleGower reviewed (approved):
Comments varying from typos to reorganize shared code.  A few of this I noted could be postponed to later ticket(s).  If so, please go ahead and make the new tickets.  There was one section of code that I noted was difficult for me to figure out.   While trying to figure it out, I was able to set up and run some tests of my own.  This means I could understand the instructions and the test cases worked.  Feel free to merge the code after fixing (or creating the new tickets).

mxk62 committed:
Fix ordering of imports

mxk62 committed:
Remove explicit inheritance from object in classes

mxk62 committed:
Fix case in constant definitions

mxk62 committed:
Use f-strings only when needed

Some of the log messages where using f-strings though plain ones could
be used instead.  Fixed it.

mxk62 committed:
Reduce code repetition

There was a lot repeated code in commands managing different components
of the manager.  I factored it out to functions to reduce the code
repetition.

mxk62 committed:
Rectify reporting issues

ktlim commented on doc/lsst.dbb.buffmngrs.endpoint/get-started.rst:
```diff
@@ -269,13 +269,80 @@ to make another ingest attempt for selected files.
+        sources:
+          - "2020-12-31"
+        search:
+          blacklist:
```
Please use a term like `exclude` rather than `blacklist`.

ktlim commented on doc/lsst.dbb.buffmngrs.endpoint/get-started.rst:
```diff
@@ -269,13 +269,80 @@ to make another ingest attempt for selected files.
+   .. note::
+
+      To prevent Python YAML parser from accidental conversion of strings like
+      ``2021-01-01`` into a ``datetime`` objects, enclose them in quotes.
```
"into a" → "into"

ktlim commented on python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py:
```diff
@@ -0,0 +1,224 @@
+        self.Event = event_creator(config["tablenames"])
+        self.File = file_creator(config["tablenames"])
+
+        self.storage = os.path.abspath(config["storage"])
```
If you are doing lots of `os.path` calls, it may be better to use `pathlib`.

ktlim commented on python/lsst/dbb/buffmngrs/endpoint/search.py:
```diff
@@ -0,0 +1,211 @@
+
+
+@register
+def scan(directory, blacklist=None, **kwargs):
```
Please change the name of the argument throughout this module.

ktlim reviewed (commented):
Apologies for not looking at this earlier, but I have a few suggestions.

mxk62 commented:
If you don't mind, I'd like to address the issues you raised as a part of the other ticket I'm currently working on (DM-28494).

ktlim commented:
Yes, that's fine.
